src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py

```diff
@@ -0,0 +1,256 @@
+        timestep: Union[float, torch.FloatTensor],
+        sample: torch.FloatTensor,
+        return_dict: bool = True,
+    ) -> Union[EulerAncestralDiscreteSchedulerOutput, Tuple]:
+        """
+        Predict the sample at the previous timestep by reversing the SDE. Core function to propagate the diffusion
+        process from the learned model outputs (most often the predicted noise).
+
+        Args:
+            model_output (`torch.FloatTensor`): direct output from learned diffusion model.
+            timestep (`float`): current timestep in the diffusion chain.
+            sample (`torch.FloatTensor`):
+                current instance of sample being created by diffusion process.
+            return_dict (`bool`): option for returning tuple rather than EulerAncestralDiscreteSchedulerOutput class
+
+        Returns:
+            [`~schedulers.scheduling_utils.EulerAncestralDiscreteSchedulerOutput`] or `tuple`:
+            [`~schedulers.scheduling_utils.EulerAncestralDiscreteSchedulerOutput`] if `return_dict` is True, otherwise
+            a `tuple`. When returning a tuple, the first element is the sample tensor.
+
+        """
+        if not self.is_scale_input_called:
+            warnings.warn(
+                "The `scale_model_input` function should be called before `step` to ensure correct denoising. "
+                "See `StableDiffusionPipeline` for a usage example."
+            )
+
+        if isinstance(timestep, torch.Tensor):
+            timestep = timestep.to(self.timesteps.device)
+        if (
+            isinstance(timestep, int)
+            or isinstance(timestep, torch.IntTensor)
+            or isinstance(timestep, torch.LongTensor)
+        ):
+            deprecate(
+                "timestep as an index",
+                "0.8.0",
+                "Passing integer indices (e.g. from `enumerate(timesteps)`) as timesteps to"
+                " `EulerAncestralDiscreteScheduler.step()` will not be supported in future versions. Make sure to pass"
+                " one of the `scheduler.timesteps` as a timestep.",
+                standard_warn=False,
+            )
+            step_index = timestep
+        else:
+            step_index = (self.timesteps == timestep).nonzero().item()
+        sigma = self.sigmas[step_index]
+
+        # 1. compute predicted original sample (x_0) from sigma-scaled predicted noise
+        pred_original_sample = sample - sigma * model_output
+        sigma_from = self.sigmas[step_index]
+        sigma_to = self.sigmas[step_index + 1]
+        sigma_up = (sigma_to**2 * (sigma_from**2 - sigma_to**2) / sigma_from**2) ** 0.5
+        sigma_down = (sigma_to**2 - sigma_up**2) ** 0.5
+
+        # 2. Convert to an ODE derivative
+        derivative = (sample - pred_original_sample) / sigma
+
+        dt = sigma_down - sigma
+
+        prev_sample = sample + derivative * dt
+
+        prev_sample = prev_sample + torch.randn_like(prev_sample) * sigma_up
```

Comment:
The euler schedulers are stochastic, and we don't pass `generator` yet in `schedulers`. Also pointed out by @tonetechnician 

Should we allow passing `generator` to `step` ? If so how would we make it generic, I could think of two ways:
-  Check if the `scheduler` accepts it, like we do it with `eta` right now 
- Add it to every `step` function, but this means it will be unused in other schedulers.

cc @patrickvonplaten @anton-l @pcuenca wdyt ?

Reply:
Let's definitely pass a generator here :-) This can be done just as is done here: https://github.com/huggingface/diffusers/blob/ea01a4c7f9ab979b00cb8cd38a2ecb409d8cf279/src/diffusers/schedulers/scheduling_ddim.py#L202

Reply:
> Hey @tonetechnician you are right, and actually that's why the tests were failing. If we pass the `generator` we need to make sure that it is generic across all schedulers as we want all schedulers in `diffusers` to be interchangeable. Let's wait to see what @patrickvonplaten @anton-l @pcuenca has to say. I have proposed two ways on the comment above.

Sounds good. I was going to ask whether there was a test to check this.

> Very good point! The cool thing about `Generator` is that, each time it's used in a random function it's state gets advanced, so if we pass `generator` we can be sure that it will be random each time.

Ah, that is super cool! Wasn't aware of that. Good to know!

Reply:
+1